Group mount partitions in the same table

dbw9580 opened the pull request:
### What changes are proposed in this pull request?

If multiple partitions in a table share the same storage location prefix, we can mount the parent table instead of mounting every partitions, thus reducing number of mount points.

### Why are the changes needed?

Fixes #13529.

### Does this PR introduce any user facing changes?

No.

This PR is based on my previous PR #13581, the last commit is the new commit to be merged in this PR.

maobaolong commented:
@dbw9580 This PR looks like base on your bypass partition branch? Would you please rebase to master branch?

maobaolong commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -148,63 +149,82 @@ public String getName() {
-              + " AlluxioLocation: " + alluxioUri
-              + " error: " + e.getMessage(), e);
+      throw new IOException(String.format(
+          "Failed to mount table location. " +
```
table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:212:46: error: '+' should be on a new line.

dbw9580 replied:
Yep. Fixed in 3f089c6.

maobaolong reviewed (commented):
A style issue, please fix it.

maobaolong reviewed (changes requested):
Thanks for the useful feature you did. After further review, please make sure this PR support group mount table in the same prefix.

A suggestion approach, we don't mount path one by one, but do a analysis and group the table in same prefix.

maobaolong commented:
@dbw9580 Could you please show some example to explain how this PR can reduce the number of mount points?

Or you can add some necessary test cases, it will also helps to let me understand the expected behavior.

dbw9580 commented:
@maobaolong The logic is:
1. The mounting happens when the UDB is created. Previously, the mounting happens when each table is requested.
2. For example, when the UDB contains 3 tables that are spread across 2 HDFS nodes:
    Table 1: `hdfs://ns1/db1/table1`
    Table 2: `hdfs://ns2/db1/table2`
    Table 3: `hdfs://ns1/db1/table3`
    We find each table's parent path, and group the common prefixes: `hdfs://ns1/db1` and `hdfs://ns2/db2`. We create 2 mount points for these 2 UFS paths on `/catalog/db1/fragments/ns1` and `/catalog/db1/fragments/ns2`, respectively. Now the tables can be accessed by appending the table's name to the URIs: `/catalog/db1/fragments/ns1/table1`, `/catalog/db1/fragments/ns2/table2` and `/catalog/db1/fragments/ns1/table3`.
3. For partitions, there are 2 cases: one, if the partition  resides in the parent table's directory, it can be found with the URI of the table, e.g. a partition of table 1 `hdfs://ns1/db1/table1/part1` is mapped to `/catalog/db1/fragments/ns1/table1/part1`, so no special handling is needed. Two, if otherwise the partition is stored somewhere else, then mount the partition specifically, and create mapping for it.

maobaolong commented:
@dbw9580 It can be more accurate to include scheme into the Alluxio mount point, so the mount point would be /catalog/db1/fragments/hdfs/ns1 and /catalog/db1/fragments/hdfs/ns12 to handle the following case

Table 1: hdfs://ns1/db1/table1
Table 2: hdfs://ns2/db1/table2
Table 3: hdfs://ns1/db1/table3
Table 4: odfs://ns1/db1/table4

dbw9580 commented:
Exactly. I left an escape method for this purpose. It's not really implemented for now, but we can discuss how it should be implemented.
https://github.com/Alluxio/alluxio/blob/57d2688f796a70e71a33b42b4641b19351e11e83/table/server/common/src/main/java/alluxio/table/common/CatalogPathUtils.java#L87-L90

maobaolong commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -146,70 +155,142 @@ public String getName() {
+    List<String> tableNames = getTableNames();
+    Map<Table, List<Partition>> tables = new HashMap<>(tableNames.size());
+    for (String tableName : tableNames) {
+      Table table;
```
If we only use the `table` in the try block, just define it in the try block

dbw9580 replied:
Fixed in 047139a5b49da41a4c8da063897dd0f39fee5098.

dbw9580 commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -196,27 +198,27 @@ private void mount(List<Table> tables, UdbBypassSpec bypassSpec)
-      AlluxioURI parentUfsUri = tableUfsUri.getParent();
-      colocatedTables.put(parentUfsUri, table);
+      AlluxioURI rootUfsUri = new AlluxioURI(tableUfsUri, "/", false);
+      colocatedTables.put(rootUfsUri, table);
```
After discussion with @maobaolong, I extended the scope of mount to the root of UFS. This is based on the assumption that if the table is accessible by Alluxio, then any leading path before the table, including the root, is accessible, too. In other words, we only need to mount the roots of every UFS to be able to access the tables contained in various locations in the UFS's, instead of finding and mounting them one by one. I tested it with a non-clustered setup of Hive and Presto, and it worked.

Would appreciate your comments on this approach, @apc999 .

dbw9580 commented on table/server/common/src/main/java/alluxio/table/common/CatalogPathUtils.java:
```diff
@@ -85,8 +85,13 @@ public static AlluxioURI getFragmentsPath(String dbName, AlluxioURI fragmentUri)
+    String escapedFragmentPath = String.format("%s_%s", scheme, authority)
+        .replaceAll(":", "_")
+        .replaceAll("/", "_");
+    return escapedFragmentPath;
```
This is to make sure we have some kind of identifier to distinguish these fragments. Since we mount the root of each UFS, the path leading to the tables is no longer relevent.

dbw9580 commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -340,7 +334,11 @@ public UdbTable getTable(String tableName, UdbBypassSpec bypassSpec) throws IOEx
       }
 
-      mount(bypassSpec);
+      if (!mIsMounted) {
+        // TODO(bowen): bypass spec updates are ignored
+        mount(bypassSpec);
+        mIsMounted = true;
+      }
```
This is problematic for now. If tables are dropped from outside of alluxio between two calls of getTable, then the path mapping info in the path translator will become stale.

maobaolong commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -146,70 +155,142 @@ public String getName() {
+    }
+    for (Map.Entry<AlluxioURI, Collection<Table>> entry : colocatedTables.asMap().entrySet()) {
+      AlluxioURI rootUfsUri = entry.getKey();
+      Collection<Table> childTables = entry.getValue();
```
It seems you defined this variable but never used it

dbw9580 replied:
Removed in 047139a5.

maobaolong reviewed (commented):
Left a couple of comments, please take a double check.

dbw9580 commented:
The lastest design of this feature is as follows:

Suppose we have 3 tables spread across two HDFS nodes:

```
Table 1: hdfs://ns1/path1/db1/table1
Table 2: hdfs://ns1/path2/db1/table2
Table 3: hdfs://ns2/path3/db1/table3
```

We create 2 mount points on `/catalog/db1/fragments/hdfs_ns1` and `/catalog/db1/fragments/hdfs_ns2`, pointing to `hdfs://ns1` and `hdfs://ns2`, respectively. We add path mappings as follows:

```
hdfs://ns1 -> /catalog/db1/fragments/hdfs_ns1
hdfs://ns2 -> /catalog/db1/fragments/hdfs_ns2
```

The tables will now be accessible from Alluxio on these Paths:

```
Table 1: /catalog/db1/fragments/hdfs_ns1/path1/db1/table1
Table 2: /catalog/db1/fragments/hdfs_ns1/path2/db1/table2
Table 3: /catalog/db1/fragments/hdfs_ns2/path3/db1/table3
```

If there are partitions in the tables, and they are not stored along with their parent tables physically, e.g. a partition `part1` of table 1 that is stored in another directory `path4` on ns1, and a partition `part2` of table 2 that is stored on another name node ns3:

```
Partition 1: hdfs://ns1/path4/db1/table1/part1
Partition 2: hdfs://ns3/path2/db1/table2/part2
```

They will be mounted individually. In summary, as long as a partition is not stored in the same directory tree of its parent table, then it will be handled as an exception and mounted specifically.

maobaolong commented on table/server/common/src/main/java/alluxio/table/common/udb/UdbContext.java:
```diff
@@ -98,4 +98,12 @@ public String getUdbDbName() {
   public AlluxioURI getTableLocation(String tableName) {
     return CatalogPathUtils.getTablePathUdb(mDbName, tableName, mType);
```
This method cannot return the right AlluxioURI?

dbw9580 replied:
This method will be deprecated once this PR gets approved, since the directory `/catalog/db/tables` will not be used any more.

maobaolong commented on table/server/common/src/main/java/alluxio/table/common/udb/UdbUtils.java:
```diff
@@ -45,23 +45,50 @@ private UdbUtils() {}
    * @throws IOException
    * @throws AlluxioException
    */
+  // Maybe rename to mountTable?
```
Remove this comment

maobaolong commented:
@dbw9580 Thanks for the improvement, @apc999 Would you please take a look at this PR, it change the catalogs mount point layout

maobaolong commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -146,72 +152,124 @@ public String getName() {
+   * @param bypassSpec bypass spec
+   */
+  private void mount(UdbBypassSpec bypassSpec) throws IOException {
+    List<String> tableNames = getTableNames();
```
This method will get all tables from udb, it don't respect the `catalog.db.ignore.udb.tables` options

dbw9580 replied:
This logic is handled by `alluxio.master.table.Database#sync` method. The ignored tables may be mounted along with others, but they will not be visible to Alluxio client.

dbw9580 replied:
Fixed in 5326f0c146e68f9d3923e2d42d2a2d3089f33d8f

apc999 commented on table/server/common/src/main/java/alluxio/table/common/udb/UdbInExClusionSpec.java:
```diff
@@ -15,44 +15,69 @@
   /**
-   * @param tablePartMap table to partition map
+   * Set of ignored tables.
+   * Tables are ignored as a whole, no partition configuration is needed.
```
then shall we name this `mIgnoredTables`

maobaolong commented on table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:
```diff
@@ -253,15 +314,19 @@ public UdbTable getTable(String tableName, UdbBypassSpec bypassSpec) throws IOEx
       }
 
-      PathTranslator pathTranslator = mountAlluxioPaths(table, partitions, bypassSpec);
+      if (!mIsMounted) {
```
This condition check is not thread-safe, suggest to move this mount logic before invoke tasks.

dbw9580 commented:
Added commits from @maobaolong 's [branch](https://github.com/maobaolong/alluxio/commits/sds-tested), and added a switch `mount.group_mount_points` to control which mounting behaviour to use.

There are open questions on the new `mount(tableNames, bypassSpec)` method of the `UnderDatabase` interface:

1. when it is called multiple times with different sets of table names, there can be new tables to mount which are not present during a previous call; tables that were mounted during a previous call, but are removed in a second call. We should skip the tables that are present in both calls, and mount the new tables, and maybe unmount the removed tables. If group mount points is enabled, there can be updates to mount points as well;
2. Updates can happen to the bypass specification between two calls to `mount`. That means some tables were be ignored before and are no longer ignored now.

dbw9580 commented:
@HelloHorizon @apc999 Rebased on current master, and added unit tests for the group mount feature. Please have a look, thanks!
